ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java

```diff
@@ -314,13 +290,20 @@ private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
             new CloudBlobMetadata(blobId, messageInfo.getOperationTimeMs(), messageInfo.getExpirationTimeInMs(),
                 messageInfo.getSize(), encryptionOrigin);
         InputStream uploadInputStream = new ByteBufferInputStream(messageBuf);
-        requestAgent.doWithRetries(() -> cloudDestination.uploadBlob(blobId, size, blobMetadata, uploadInputStream),
-            "Upload", partitionId.toPathString());
+        uploaded =
+            requestAgent.doWithRetries(() -> cloudDestination.uploadBlob(blobId, size, blobMetadata, uploadInputStream),
+                "Upload", partitionId.toPathString());
       }
       addToCache(blobId.getID(), (short) 0, BlobState.CREATED);
+      if (!uploaded) {
+        throw new StoreException(String.format("Another blob with same key %s exists in store", blobId.getID()),
+            StoreErrorCodes.Already_Exist);
+      }
     } else {
-      logger.trace("Blob is skipped: {}", messageInfo);
       vcrMetrics.blobUploadSkippedCount.inc();
+      throw new CloudStorageException("Error updating blob metadata", new StoreException(
```

Comment:
Why not just throw a StoreException here?
Also, for VCR case it may not be an error (that's why we mark it skipped).